Fix conversion of Zarr string dataset to HDF5

oruebel opened the pull request:
## Motivation

Fix https://github.com/hdmf-dev/hdmf-zarr/issues/211

Zarr does not natively support variable length strings, instead the data is stored as objects with an encoder. When converting from Zarr to HDF5 the `ObjectMapper` tries to change the dtype in some cases for text datasets to unicode, which in turn can cause reading the data from Zarr to fail. This PR changes this behavior to: 1) not change the dtype for Zarr string datasets, and 2) detect the correct dtype to use in HDF5 in the backend. 

## Checklist

- [x] Did you update `CHANGELOG.md` with your changes?
- [X] Does the PR clearly describe the problem and the solution?
- [X] Have you reviewed our [Contributing Guide](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst)?
- [X] Does the PR use "Fix #XXX" notation to tell GitHub to close the relevant issue numbered XXX when the PR is merged?

oruebel committed:
Fix conversion of Zarr string dataset to HDF5

oruebel committed:
Updated changelog

oruebel committed:
Fix spelling

oruebel commented:
@rly could you please take a look to see if this is right way to address this issue

rly committed:
Merge branch 'dev' into fix/zarr_str_export

rly commented on src/hdmf/backends/hdf5/h5tools.py:
```diff
@@ -923,6 +923,12 @@ def __resolve_dtype__(cls, dtype, data):
+
+        # Use text dtype for Zarr datasets of strings. Zarr stores variable length strings
+        # as objects, so we need to detect this special case here
+        if hasattr(data, 'attrs') and 'zarr_dtype' in data.attrs and data.attrs['zarr_dtype'] == 'str':
+            return cls.__dtypes['text']
```
If we add:
```python
try:
    from zarr import Array as ZarrArray
    import numcodecs
    ZARR_INSTALLED = True
except ImportError:
    ZARR_INSTALLED = False

def _is_zarr_array(value):
    return ZARR_INSTALLED and isinstance(value, ZarrArray)
```

Can we change these lines to:
```python
        if _is_zarr_array(data) and data.filters:
            if numcodecs.VLenUTF8() in data.filters:
                return cls.__dtypes['text']
            elif numcodecs.VLenBytes() in data.filters:
                return cls.__dtypes['ascii']
```

It would be nice if `HDF5IO` can accept any Zarr array just like it accepts any Numpy array, instead of accepting only Zarr arrays that have been written with hdmf-zarr. The Zarr array filters should show that the array dtype is a variable-length string and its encoding.

rly replied:
That said, I don't seem to be able to create a test case where these lines (regardless of yours or mine) matters. Could you add one?

rly replied:
Note from discussion: this probably can be removed and was there to be overly cautious. Ryan TODO: try a conversion and confirm.

rly committed:
Modification to Fix conversion of Zarr string dataset to HDF5 (#1175)

rly committed:
Merge branch 'dev' into fix/zarr_str_export

rly committed:
Fix test

rly commented:
This PR also resolves this issue: https://github.com/hdmf-dev/hdmf-zarr/issues/301

@oruebel could you please review?

rly committed:
Update CHANGELOG.md

Added upcoming release notes for HDMF 4.1.2 and fixed formatting issues in the changelog.

oruebel commented:
@rly thanks for updating the PR. This looks good to me.
